lasagne/layers/merge.py

```diff
@@ -327,14 +328,35 @@ def __init__(self, incomings, merge_function, cropping=None, **kwargs):
 
     def get_output_shape_for(self, input_shapes):
         input_shapes = autocrop_array_shapes(input_shapes, self.cropping)
+
+        input_dims = [len(shp) for shp in input_shapes]
+        if not all(input_dim == input_dims[0] for input_dim in input_dims):
+            raise ValueError('Input dimensions must be the same but were %s' %
+                             ", ".join(map(str, input_shapes)))
+
+        def broadcasting(input_dim):
+            # handle the broadcastable case where 1's are broadcasted
+            sorted_dim = sorted([-1 if d is None else d for d in input_dim])
+            if isinstance(sorted_dim[-1], int) and sorted_dim[-1] > 0 and all(
+                    [d == 1 or d is -1 for d in sorted_dim[:-1]]):
+                broadcast_size = sorted_dim[-1]
+                return [broadcast_size]*len(input_dim)
+            else:
+                return input_dim
+
+        # if dimension is broadcastable and we can infer the size then then
+        # replace all sizes with the broadcasting dimension.
+        input_dims = [broadcasting(input_dim)
+                      for input_dim in zip(*input_shapes)]
+        input_shapes = zip(*input_dims)
+
```

Comment:
This is a lot of code. Can't you just change the `output_shape` computation to grab the largest input size per axis instead of the first, and then change the `match` code to treat `1` like `None`? As far as I see, you won't need to add any more code then.